### PR TITLE
tree table snapshots

### PR DESCRIPTION
**Pull request by spasovski**
missing formatting and I don't think the intersection is quite right but it's quite the snack

### CODE REVIEW

File: packages/iris-grid/src/IrisGridTreeTableModel.js [hunk 1]
```diff
@@ -32,6 +37,35 @@ class IrisGridTreeTableModel extends IrisGridTableModel {
...
+    const viewportRange = new GridRange(
+      0,
+      this.viewportData.offset,
+      columns.length,
+      this.viewportData.offset + this.viewportData.rows.length
+    );
+    const intersection = GridRange.intersection(viewportRange, ranges);
+
+    for (let i = intersection.startRow; i < intersection.endRow; i += 1) {
+      const row = this.viewportData.rows[i];
```
Comment (mofojed) [inline packages/iris-grid/src/IrisGridTreeTableModel.js:52]: You'll need to iterate through the columns as well, e.g.
```
  const resultRow = [];
  const viewportRow = this.viewportData.rows[i - this.viewportData.offset];
  for (let j = intersection.startColumn; j <= intersection.endColumn; j += 1) {
    resultRow.push(formatValue(viewportRow.data.get(j).value, this.columns[j]));
  }
  result.push(resultRow);
}
```

File: packages/iris-grid/src/IrisGridTreeTableModel.js [hunk 2]
```diff
@@ -32,6 +37,35 @@ class IrisGridTreeTableModel extends IrisGridTableModel {
...
+    const result = [];
+    const viewportRange = new GridRange(
+      0,
+      this.viewportData.offset,
+      columns.length,
+      this.viewportData.offset + this.viewportData.rows.length
+    );
+    const intersection = GridRange.intersection(viewportRange, ranges);
+
+    for (let i = intersection.startRow; i < intersection.endRow; i += 1) {
```
Comment (mofojed) [inline packages/iris-grid/src/IrisGridTreeTableModel.js:51]: Needs to be `i <= intersection.endRow`, as the index is inclusive

File: packages/iris-grid/src/IrisGridTreeTableModel.js [hunk 3]
```diff
@@ -32,6 +37,35 @@ class IrisGridTreeTableModel extends IrisGridTableModel {
     };
   }
 
+  snapshot(ranges) {
```
Comment (mofojed) [inline packages/iris-grid/src/IrisGridTreeTableModel.js:40]: This still needs to be async, it needs to match the signature of the super IrisGridTableModel
Side note IrisGridModel should update the signature to match IrisGridTableModel as well

File: packages/iris-grid/src/IrisGridTreeTableModel.js [hunk 4]
```diff
@@ -32,6 +37,35 @@ class IrisGridTreeTableModel extends IrisGridTableModel {
...
+    const intersection = GridRange.intersection(viewportRange, ranges);
+
+    for (let i = intersection.startRow; i < intersection.endRow; i += 1) {
+      const row = this.viewportData.rows[i];
+      result.push([...row.data.values()].map(rowVal => rowVal.value));
+    }
+    return result;
+  }
+
+  textSnapshot(
```
Comment (mofojed) [inline packages/iris-grid/src/IrisGridTreeTableModel.js:58]: I don't think you need to override `textSnapshot` if you're override `snapshot`

**Review by mofojed — CHANGES_REQUESTED** [review]

File: packages/iris-grid/src/IrisGridTreeTableModel.js [hunk 5]
```diff
@@ -32,6 +33,40 @@ class IrisGridTreeTableModel extends IrisGridTableModel {
...
+
+    for (let i = 0; i < ranges.length; i += 1) {
+      const intersection = GridRange.intersection(viewportRange, ranges[i]);
+
+      for (let r = intersection.startRow; r <= intersection.endRow; r += 1) {
+        const resultRow = [];
+        const viewportRow = this.viewportData.rows[
+          r - this.viewportData.offset
+        ];
+        for (
```
Comment (spasovski) [inline packages/iris-grid/src/IrisGridTreeTableModel.js:54]: It's linear time just trust me...

**Review by mofojed — CHANGES_REQUESTED** [review]

**Review by mofojed — APPROVED** [review]

**Pull request merged**